fix: envoy service cluster name for zone-aware routing

y-rabie opened the pull request:
**What type of PR is this?**

fix!: fix envoy service cluster name for zone-aware routing

**What this PR does / why we need it**:

Currently, the new zone aware routing is not always functional when not using ForceLocalZone and using Envoy's normal distributional zone-aware routing.

This is the case when having a hard-coded name for the envoy's service in EnvoyProxy. The code always assumes the service has an autogenerated name, and so in other cases the discovery fails for envoy endpoints.

I've moved the call for `processServiceCluster` in case of unmerged gateways down after we do the `processGatewayParamsRef`, so that we can fetch EnvoyProxy directly from the `resourceTree`. If this might cause any other unintended problems, I can move it to where it was, and fetch EnvoyProxy manually instead.

Fixes #

<!--
For any non-trivial changes, you need to provide a brief description of the changes in the release notes.
Please add the description to the release-notes/current.yaml file and include this file in the PR.
-->
Release Notes: Yes/No

jukie commented:
Apologies for the oversight. I'll review this!

jukie commented:
Thanks! Could you add an internal test case or another zone aware e2e in https://github.com/envoyproxy/gateway/blob/c2404fedaad0fb2a3e19c8d8aaac23f9d3df5c4c/test/e2e/tests/zone_aware_routing.go?

zirain commented on internal/provider/kubernetes/controller.go:
```diff
@@ -1487,13 +1493,22 @@ func (r *gatewayAPIReconciler) processGateways(ctx context.Context, managedGC *g
 		return
 	}
 
 	proxySvcName := proxy.ExpectedResourceHashedName(resourceName)
```
we need to check if gateway mode is enabled.

y-rabie replied:
Can you clarify what is "gateway mode" specifically?

zirain replied:
you can find the detail here: https://gateway.envoyproxy.io/docs/tasks/operations/gateway-namespace-mode/

In short, deployment/svc will be provisioned the namespace with Gateway resource.

y-rabie replied:
Fixed, PTAL

y-rabie commented:
@jukie Adjusted tests. PTAL and re-run them

zirain commented on internal/provider/kubernetes/controller.go:
```diff
@@ -1487,16 +1493,31 @@ func (r *gatewayAPIReconciler) processGateways(ctx context.Context, managedGC *g
+	}
+
+	// Check if gateway namespaced mode is used
+	proxySvcNamespace := r.namespace
```
move L1513-L1517 before L1505?

jukie commented on test/e2e/testdata/zone-aware-routing-deployments.yaml:
```diff
@@ -57,7 +57,6 @@ metadata:
   labels:
     app: zone-aware-backend
 spec:
-  replicas: 1
```
Is there a reason to remove this?

zirain commented on internal/provider/kubernetes/controller.go:
```diff
@@ -1487,16 +1493,30 @@ func (r *gatewayAPIReconciler) processGateways(ctx context.Context, managedGC *g
 }
 
-func (r *gatewayAPIReconciler) processServiceCluster(resourceName string, resourceMap *resourceMappings) {
+func (r *gatewayAPIReconciler) processServiceCluster(ep *egv1a1.EnvoyProxy, resourceNamespacedName types.NamespacedName, resourceMap *resourceMappings) {
```
not a block, we should add a test somewhere to make sure that this will not be broken again.

y-rabie commented:
I've split the function into two, one for GatewayClass and one for Gateway. There are subtle differences between the two, and tbh having them in one polymorphic function would be less clear. 

I've also added unit tests for both, and e2e tests should be successful now.

jukie commented:
/retest

jukie commented:
/retest
